src/ocf_stats_builder.c

```diff
@@ -346,32 +349,37 @@ int ocf_stats_collect_core(ocf_core_t core,
 
 	if (usage) {
 		_set(&usage->occupancy,
-			_lines4k(s.cache_occupancy, cache_line_size),
+			_lines4k(s->cache_occupancy, cache_line_size),
 			_lines4k(cache_size, cache_line_size));
 
 		_set(&usage->free,
 			_lines4k(cache_size - cache_occupancy, cache_line_size),
 			_lines4k(cache_size, cache_line_size));
 
 		_set(&usage->clean,
-			_lines4k(s.cache_occupancy - s.dirty, cache_line_size),
-			_lines4k(s.cache_occupancy, cache_line_size));
+			_lines4k(s->cache_occupancy - s->dirty, cache_line_size),
+			_lines4k(s->cache_occupancy, cache_line_size));
 
 		_set(&usage->dirty,
-			_lines4k(s.dirty, cache_line_size),
-			_lines4k(s.cache_occupancy, cache_line_size));
+			_lines4k(s->dirty, cache_line_size),
+			_lines4k(s->cache_occupancy, cache_line_size));
 	}
 
 	if (req)
-		_fill_req(req, &s);
+		_fill_req(req, s);
 
 	if (blocks)
-		_fill_blocks(blocks, &s);
+		_fill_blocks(blocks, s);
 
 	if (errors)
-		_fill_errors(errors, &s);
-
+		_fill_errors(errors, s);
+	
+	env_vfree(s)
 	return 0;
+
+collect_failed:
+	env_vfree(s);
+	return result;
```

Comment:
Actually, since 'result' must equal 0 at all times for this function to complete successfully, I would remove the first free and return statements and leave only those after 'collect_failed' mark.

Reply:
And you may also rename collect_failed to someting like mem_free maybe? This will be less confusing during normal non-error flow.

Reply:
Good point It was in my mind for a moment. Will be changed :)